Apply select preprocessor to area, volume and zonal statistics fx variables and add use of fx_files as list of dictionaries

valeriupredoi opened the pull request:
Before you start, please read [CONTRIBUTING.md](https://github.com/ESMValGroup/ESMValCore/blob/master/CONTRIBUTING.md).

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [ ] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #436 #440

valeriupredoi commented:
OK @ledm this is now functional, can you give it a test please? To derive fx vars for `volume_statistics` add `fx_var_preprocess: True` in variable eg:
```
        diagnostics:
          diagnostic_name:
            variables:
              tos:
                preprocessor: preproc
                fx_var_preprocess: True
                project: CMIP6
                mip: Omon
                exp: historical
                start_year: 2000
                end_year: 2005
                ensemble: r1i1p1f1
                grid: gn
                additional_datasets:
                  - {dataset: CanESM5}
```

ledm commented on esmvalcore/_recipe.py:
```diff
@@ -890,6 +910,34 @@ def _get_preprocessor_task(variables, profiles, config_user, task_name):
             derive_tasks.append(task)
 
+    # special case: fx variable pre-processing
+    for step in ('area_statistics', 'volume_statistics'):
```
I think `zonal_statistics` should be added to this list.
```suggestion
    for step in ('area_statistics', 'volume_statistics', 'zonal_statistics'):
```

ledm commented on esmvalcore/_recipe.py:
```diff
@@ -489,12 +494,27 @@ def _update_fx_settings(settings, variable, config_user):
         logger.info(msg, 'land/sea fraction weighting', pformat(fx_dict))
 
     for step in ('area_statistics', 'volume_statistics'):
```
I think `zonal_statistics` should be added to this list.
```suggestion
    for step in ('area_statistics', 'volume_statistics', 'zonal_statistics'):
```

mattiarighi replied:
`meridional_statistics` as well?

valeriupredoi commented:
aye, finally! This was tested by me with a (rather comperehensive ) recipe:
```
# recipe_fxprep.yml
---
documentation:

  description: test recipe for fx var preprocessing for area/volume/zonal/statistics

  authors:
    - predoi_valeriu

preprocessors:
    prep:
        custom_order: true
        extract_volume:
            z_min: 0
            z_max: 100
        annual_statistics:
            operator: mean
        volume_statistics:
            operator: mean
            fx_files: [volcello, ]

diagnostics:
  diag:
    variables:
      thetao700_Omon:
        short_name: thetao
        preprocessor: prep
        fx_var_preprocess: True
        additional_datasets:
          - {dataset: GFDL-CM4,        project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f1, grid: gn, start_year: 2010, end_year: 2014}
          - {dataset: HadGEM3-GC31-LL, project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f3, grid: gn, start_year: 2010, end_year: 2014}
          - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: historical, ensemble: r2i1p1f2, grid: gn, start_year: 2010, end_year: 2014}
```
@mattiarighi and @ledm pls fire up tests yourselves. There may still be corner cases that I haven't looked at but since that recipe above works am quite happy :beer: I need to add documentation!

valeriupredoi commented:
OK chaps, this is a bit involved, and I am going to summarize in text (that I will also add to the doc) what this does; before I do that I'd like to send a special request to my boi @schlunma to have a look at the PR since some of the changes are a direct followup on his implementation from last week :beer:

New features:
 - it is possible to run a set of preprocessor steps on the `fx_files: [vars or dicts of vars]` specified in the `area_statistics`, `volume_statistics` or `zonal_statistics`by setting the `fx_var_preprocess: True` in the `variable`; take for instance this example:
```
preprocessors:
    prep:
        custom_order: true
        extract_volume:
            z_min: 0
            z_max: 100
        annual_statistics:
            operator: mean
        volume_statistics:
            operator: mean
            fx_files: [volcello, ]

diagnostics:
  diag:
    variables:
      thetao700_Omon:
        short_name: thetao
        preprocessor: prep
        fx_var_preprocess: True
        additional_datasets:
          - {dataset: GFDL-CM4,        project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f1, grid: gn, start_year: 2010, end_year: 2014}
          - {dataset: HadGEM3-GC31-LL, project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f3, grid: gn, start_year: 2010, end_year: 2014}
          - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: historical, ensemble: r2i1p1f2, grid: gn, start_year: 2010, end_year: 2014}
    scripts:
      null
```
this setup will run the custom-ordered preprocessor steps in `preprocessors/prep` before (and not including) `volume_statistics` for the fx variable `volcello`;

- it is also possible to request the same type of fx variables but as dictionaries, providing more information towards the retrieval of the correct data; take this snippet for example:
```
preprocessors:
    prep:
        custom_order: true
        extract_volume:
            z_min: 0
            z_max: 100
        annual_statistics:
            operator: mean
        volume_statistics:
            operator: mean
            fx_files: [{short_name: volcello, mip: Omon, exp: historical}]

diagnostics:
  diag:
    variables:
      thetao700_Omon:
        short_name: thetao
        preprocessor: prep
        fx_var_preprocess: False
        additional_datasets:
          - {dataset: GFDL-CM4,        project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f1, grid: gn, start_year: 2010, end_year: 2014}
          - {dataset: HadGEM3-GC31-LL, project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f3, grid: gn, start_year: 2010, end_year: 2014}
          - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: historical, ensemble: r2i1p1f2, grid: gn, start_year: 2010, end_year: 2014}
    scripts:
      null
```
in this case, by providing the extra CMOR keys, we allow the code to find the `volcello` with mip `Omon` and experiment `historical`. This is particularly useful when the actual fx data is not found in the master variable's (`thetao` in this case) data experiment; this is also useful to force a certain time-dependent data (mip: Omon) for preprocessors that need time operations

ledm commented on esmvalcore/_recipe.py:
```diff
@@ -387,33 +386,54 @@ def _add_fxvar_keys(fx_var_dict, variable):
+                logger.debug("Found fx variables '%s':\n%s", fx_varname,
+                             pformat(fx_files))
                 break
     else:
```
Is this `else` statementlinked to a specific `if`? Looks like its preceded by the `for` loop?

valeriupredoi replied:
it is - the `break` case just on top of it - by @schlunma but I actually changed the declaration of `fx_files` so that it's a bit more straightforward

schlunma replied:
It's syntactic sugar of python, see e.g. https://book.pythontips.com/en/latest/for_-_else.html

valeriupredoi replied:
indeed, I killed the sugar tho :grin: - should be ok no?

ledm commented on esmvalcore/_recipe.py:
```diff
@@ -890,6 +967,36 @@ def _get_preprocessor_task(variables, profiles, config_user, task_name):
+                    for fx_var in fx_vars
+                ]
+                for fx_variable in fx_variables:
+                    fx_name = task_name.split(
```
I think we need some kind of raise here for when the fx file is not found. 

At the moment this is flagged as a list index problem. If no fx files are found, then `fx_variable` is a empty list and `fx_variable['variable_group']` on line 996 just raises the error:
```
  File "/users/modellers/ledm/workspace/ESMValTool_AR6/ESMValCore/esmvalcore/_recipe.py", line 996, in _get_preprocessor_task
    fx_variable['variable_group']
TypeError: list indices must be integers or slices, not str


```

ledm replied:
This error can be reproduced with the recipe:

```
# ESMValTool
# test_439.yml
---
documentation:
  description: |
    Recipe to test PR 439

  authors:
    - demora_lee

  projects:
    - ukesm

preprocessors:
    prep_sst:
        custom_order: True
        annual_statistics:
            operator: mean
        extract_region:
            start_longitude: -80.
            end_longitude: 30.
            start_latitude: -80.
            end_latitude: 80.
        area_statistics:
            operator: mean
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl}]
        regrid_time:
            frequency: yr

diagnostics:
  diag_439:
      variables:
        thetaosst:
          short_name: tos
          preprocessor: prep_sst
          fx_var_preprocess: True
          additional_datasets:
            - {dataset: GFDL-CM4,        project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f1, grid: gn, start_year: 1960, end_year: 1963}

      scripts:
        diagnostic_439:
          script: ocean/diagnostic_timeseries.py
          #script: ocean/diagnostic_fig_3_20_OHC_noppderive_timeseries.py

```

ledm replied:
Note that this recipe will always fail, as there is no `GFDL-CM4 areacello` data available o  jasmin.

ledm commented:
At the moment, this works very well for the volume calculation. so that's great, thanks a lot V!

However, I'm also testing it with the area calculation with a recipe like this:
```
preprocessors:
    prep_sst:
        custom_order: True
        annual_statistics:
            operator: mean
        extract_region:
            start_longitude: -80.
            end_longitude: 30.
            start_latitude: -80.
            end_latitude: 80.
        area_statistics:
            operator: mean
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl}]
        regrid_time:
            frequency: yr

diagnostics:
  diag_439:
      variables:
        thetaosst:
          short_name: tos
          preprocessor: prep_sst
          fx_var_preprocess: True
          additional_datasets:
            - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: historical, ensemble: r1i1p1f2, grid: gn, start_year: 1960, end_year: 1963}

      scripts:
        diagnostic_439:
          script: ocean/diagnostic_timeseries.py

```

It's trying to apply the `annual_statistics` preprocessor to the `areacello` fx file, which does not have a time component. 

(The `fx_files` FX dictionary part works great though! It's amazing to have that level of control on finding the FX files - this has been such a pain so far!
```
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl}]
```
Thanks buddy! We just need a clear error message when it doesn't find it!)

valeriupredoi commented:
Cheers for testing @ledm :beer: but the fail is obv coz Ofx doesnt have a time coord, hence one of the uses of fxfiles as dict - to force a time varying mip like Omon. If there's no such data then time ops can not be run

valeriupredoi commented:
OK so the case where time preprocessor steps are applied (and fail!) on fx vars with *fx mips is now solved by simply removing those time-preproc steps from the preprocessor chain *only* for those fx variables - I am not entirely happy with adding a list of `TIME_PREPROCESSORS` in `preprocessor/__init__.py` but I don't see another way how to get around this in an elegant way - it may actually prove to be useful for other situations as well!

ledm commented on esmvalcore/_recipe.py:
```diff
@@ -990,13 +990,13 @@ def _get_preprocessor_task(variables, profiles, config_user, task_name):
+                    # list may be (intentionally) empty - catch it here
+                    if not fx_variable:
+                        raise RecipeError(
+                            f"One or more of {fx_vars} are missing data. "
```
This is still not a useful error message. can we change it to:

```suggestion
                            f"One or more of {step} fx data for of {var['short_name']} are missing. "
```

ledm commented on esmvalcore/_recipe.py:
```diff
@@ -990,13 +990,13 @@ def _get_preprocessor_task(variables, profiles, config_user, task_name):
+                        raise RecipeError(
+                            f"One or more of {fx_vars} are missing data. "
+                            f"Task can not be performed since there is "
+                            f"no data available. ")
```
```suggestion
                            f"no fx data found.")
```

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -387,37 +386,56 @@ def _add_fxvar_keys(fx_var_dict, variable):
-    # TODO: allow user to specify certain mip if desired
+    # first see if we have strings or dicts for fx_varname
+    user_defined = False
+    if isinstance(fx_varname, dict):
```
good call, mate, done it!

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -428,8 +446,10 @@ def _get_correct_fx_file(variable, fx_varname, config_user):
     # allow for empty lists corrected for by NE masks
     if fx_files:
         fx_files = fx_files[0]
```
yes that's a good question - normally yes, one single file, but this will fail miserably when we actually find two or more files. Unfortunately can't do anything about this at this stage, we'll correct it in a different PR if we see otherwise

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -488,13 +510,66 @@ def _update_fx_settings(settings, variable, config_user):
+                fx_file, cmor_fx_var = _get_correct_fx_file(
+                    variable, fxvar, config_user)
+                fx_files_dict[fxvar_name] = fx_file
+                if cmor_fx_var:
```
`cmor_fx_var` is either an empty list or a dict, so we checking if it is not an empty list

ledm commented:
I'm findin that having a working fx file interface (finally! thanks @valeriupredoi and @schlunma) has uncovered some issues with `_area.py`, notably the stuff to do with tiling arrays. 

Can we add those fixes in with this PR or should we do them elsewhere?

valeriupredoi commented:
> I'm findin that having a working fx file interface (finally! thanks @valeriupredoi and @schlunma) has uncovered some issues with `_area.py`, notably the stuff to do with tiling arrays.
> 
> Can we add those fixes in with this PR or should we do them elsewhere?

different PR please, this is already way bloated, plus it's a different topic :beer:

ledm commented:
Okay, but know that this PR breaks `area_statistics`.

schlunma commented on esmvalcore/_recipe.py:
```diff
@@ -387,37 +386,57 @@ def _add_fxvar_keys(fx_var_dict, variable):
+                'mip': fx_mip
+            }, var)
+            if user_fx_experiment:
+                fx_var['exp'] = user_fx_experiment
```
I think replacing all this by

```python
fx_var = _add_fxvar_keys(fx_varname, var)
```

and simplifying `_add_fxvar_keys` by

```python
def _add_fxvar_keys(fx_var_dict, variable):
    """Add keys specific to fx variable to use get_input_filelist."""
    fx_variable = dict(variable)
    fx_variable.update(fx_var_dict)
    fx_variable['variable_group'] = fx_var_dict['short_name']
    if fx_variable['project'] == 'CMIP5':
        fx_variable['ensemble'] = 'r0i0p0'
    _add_cmor_info(fx_variable, override=True)
return fx_variable
```

should also the job + allows using other options than `exp` and `mip` in the `dict` for fx_files. You can also remove l.394 to 396 then.

valeriupredoi replied:
no that's not working(tested until now, since you gave me the suggestion) because there's the mips that need to be changed for the fx variables otherwise the cmor table loading fails

schlunma replied:
Ah my bad, sorry. You need to add the `mip` to the `fx_varname` dictionary beforehand, e.g.

```python
fx_varname['mip'] = fx_mip
fx_var = _add_fxvar_keys(fx_varname, var)
```

then it should work.

valeriupredoi replied:
it does work but the last test fails:
```
def test_fx_var_invalid_project(tmp_path, patched_datafinder, config_user):
    content = dedent("""
        preprocessors:
          preproc:
           area_statistics:
             operator: mean
             fx_files: ['areacella']

        diagnostics:
          diagnostic_name:
            variables:
              tas:
                preprocessor: preproc
                project: EMAC
                mip: Amon
                exp: historical
                start_year: 2000
                end_year: 2005
                ensemble: r1i1p1f1
                grid: gn
                additional_datasets:
                  - {dataset: CanESM5}
            scripts: null
        """)
    msg = 'Project EMAC not supported with fx variables'
    with pytest.raises(RecipeError) as rec_err_exp:
        get_recipe(tmp_path, content, config_user)
        assert msg in rec_err_exp
```
I cant see anywhere the expected error - where is it?

valeriupredoi replied:
I removed the test since it was dodgy, pls put it back in (if in a working condition) but I dont see a need for it tho

valeriupredoi replied:
cool dis fixed now

ledm commented:
> > I'm findin that having a working fx file interface (finally! thanks @valeriupredoi and @schlunma) has uncovered some issues with `_area.py`, notably the stuff to do with tiling arrays.
> > Can we add those fixes in with this PR or should we do them elsewhere?
> 
> different PR please, this is already way bloated, plus it's a different topic

Created issue #445

ledm commented:
Okay, found another issue with the fx_files dictionary in the preprocessor recipe (sorry!). I mean this part:
```
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl}]
```
This seems to work fine for the `historical` experiments, as  `historical` and `piControl` (where the FX files live) are both in in the `activity` domain (`CMIP`). When we're looking at things from a different activity (like the experiment `ssp585` in `activity` `ScenarioMIP`), it can't find the FX files anymore. 

It's looking for the areacello here:
```
/badc/cmip6/data/CMIP6/ScenarioMIP/MOHC/UKESM1-0-LL/piControl/r1i1p1f2/Ofx/areacello/gn/{latestversion}
```
but that area doesn't exist, because it should be looking in the location:
```
/badc/cmip6/data/CMIP6/CMIP/MOHC/UKESM1-0-LL/piControl/r1i1p1f2/Ofx/areacello/gn/{latestversion}

```
So, I guess that we need to add the `activity` tag to the preprocessor fx dictionary, like this:
```
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl, activity: CMIP}]
```

To test this, try this recipe:
```
preprocessors:
    prep_sst:
        custom_order: True
        annual_statistics:
            operator: mean
        extract_region:
            start_longitude: -80.
            end_longitude: 30.
            start_latitude: -80.
            end_latitude: 80.
        area_statistics:
            operator: mean
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl, activity: CMIP}]
        regrid_time:
            frequency: yr

diagnostics:
  diag_439:
      variables:
        thetaosst:
          short_name: tos
          preprocessor: prep_sst
          fx_var_preprocess: True
          additional_datasets:
            - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: ssp585, ensemble: r1i1p1f2, grid: gn, start_year: 2060, end_year: 2063}

      scripts:
        diagnostic_439:
          script: ocean/diagnostic_timeseries.py
```

valeriupredoi commented:
@ledm you should be fine now

valeriupredoi commented:
just tested that recipe and it worked pronto :grin:

ledm commented:
That fixed it, but got a new error:
```
Traceback (most recent call last):
  File "/home/users/ldemora/workspace/ESMValTool_private/ESMValCore/esmvalcore/_main.py", line 220, in run
    conf = main(args)
  File "/home/users/ldemora/workspace/ESMValTool_private/ESMValCore/esmvalcore/_main.py", line 156, in main
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "/home/users/ldemora/workspace/ESMValTool_private/ESMValCore/esmvalcore/_main.py", line 202, in process_recipe
    recipe = read_recipe_file(recipe_file, config_user)
  File "/home/users/ldemora/workspace/ESMValTool_private/ESMValCore/esmvalcore/_recipe.py", line 73, in read_recipe_file
    recipe_file=filename)
  File "/home/users/ldemora/workspace/ESMValTool_private/ESMValCore/esmvalcore/_recipe.py", line 1015, in __init__
    self.tasks = self.initialize_tasks() if initialize_tasks else None
  File "/home/users/ldemora/workspace/ESMValTool_private/ESMValCore/esmvalcore/_recipe.py", line 1384, in initialize_tasks
    check.tasks_valid(tasks)
  File "/home/users/ldemora/workspace/ESMValTool_private/ESMValCore/esmvalcore/_recipe_checks.py", line 125, in tasks_valid
    raise ValueError(msg.format(product.filename))
ValueError: Duplicate preprocessor filename /home/users/ldemora/workspace/ESMValTool_private/run/esmvaltool_output/recipe_gwt_time_series_20200123_170329/preproc/time_series/areacello/CMIP6_UKESM1-0-LL_Ofx_piControl_r1i1p1f2_areacello.nc, please file a bug report.
2020-01-23 17:03:47,379 UTC [67417] INFO    If you suspect this is a bug or need help, please open an issue on https://github.com/ESMValGroup/ESMValTool/issues and attach the run/recipe_*.yml and run/main_log_debug.txt files from the output directory.
```

Looks like it's creating every preprocessed fx file with the same name!

This recipe should reproduce the error:

```
preprocessors:
    prep_sst:
        custom_order: True
        annual_statistics:
            operator: mean
        extract_region:
            start_longitude: -80.
            end_longitude: 30.
            start_latitude: -80.
            end_latitude: 80.
        area_statistics:
            operator: mean
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl, activity: CMIP}]
        regrid_time:
            frequency: yr

diagnostics:
  diag_439:
      variables:
        thetaosst:
          short_name: tos
          preprocessor: prep_sst
          fx_var_preprocess: True
          additional_datasets:
            - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: ssp585, ensemble: r1i1p1f2, grid: gn, start_year: 2060, end_year: 2063}
            - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: ssp119, ensemble: r1i1p1f2, grid: gn, start_year: 2060, end_year: 2063}

      scripts:
        diagnostic_439:
          script: ocean/diagnostic_timeseries.py
```

valeriupredoi commented on esmvalcore/_recipe.py:
```diff
@@ -488,13 +510,66 @@ def _update_fx_settings(settings, variable, config_user):
+                    cmor_fxvars.append(cmor_fx_var)
+
+            # special case for preprocessing fx variables
+            if var.get("fx_var_preprocess"):
```
I need that to be set as a special case because of this line: l.753 `if not ancestors or variable.get("fx_var_preprocess"):` -> this is a special case that allows the code to get the files for the main variable even if there are ancestors. If we set that True all the time it'll wreak havoc when there are no fx_files preprocs

valeriupredoi commented:
of course it's the same name if you are forcing the same project and the fx file is the same for two identical datasets. Hang on, I'll see what's to be done here

ledm commented:
> of course it's the same name if you are forcing the same project and the fx file is the same for two identical datasets. Hang on, I'll see what's to be done here

Cheers!

Thinking about this: Do we think we need to make one file for each dataset, or can we reuse the same preprocessed FX file here?

I think we might be able to re-use the same file. If an preprocessor fx file uses a different preprocessor, experiment or time range, it would have a different final path, right?

valeriupredoi commented:
Yea that's the idea - on bus to work and I *think* I knows how to fix this elegantly. In the meantime - I've plugged in the simplifications suggested by @schlunma with the added benefit of solving different activities (as per your previous fail) but am still puzzled abt that EMAC test (which I temporarily removed) - Manuel, can you have a look at that when you gots time, please mate :beer:

schlunma commented:
You removed this error message in e1cd96832ea4d2d6e9c13b1ea5cbd61cdbe93e3d, but I have no idea why the test did not fail in #432. It must fail for the current master branch, too...

valeriupredoi commented:
Eugh, my bad then, cheers amigo, will put it back in then complete with its message :beer:

schlunma commented:
I think this error is not necessary anymore, so you can leave it just like it is now :grin:

valeriupredoi commented:
ach so, musik to me ears :beer:

valeriupredoi commented:
OK am done, ready for the guillotine :fr:

ledm commented:
Well that worked! Excelling job V, Manuel. Thanks!
![MultipleModels__rh_prep_timeseries_2d_land_time_series_2015_2100_timeseries_](https://user-images.githubusercontent.com/4180685/73075367-a7974c00-3eb3-11ea-95b6-b296ad3cac33.png)

mattiarighi commented on doc/esmvalcore/preprocessor.rst:
```diff
@@ -414,6 +414,92 @@ variable items e.g.:
 .. _masking of missing values:
 
+
+Using mask files for area/volume/zonal statistics
```
Can you move this section after "Missing values masks"?

valeriupredoi replied:
yep on it

valeriupredoi commented:
@ledm cheers a bunch for testing, dude! Can you pls attach the recipe you are running (the more involved one) so @mattiarighi can run a test too? :beer:

ledm commented:
> @ledm cheers a bunch for testing, dude! Can you pls attach the recipe you are running (the more involved one) so @mattiarighi can run a test too?


It's basically just this one, but with a couple extra SSP experiments and longer time ranges.

https://github.com/ESMValGroup/ESMValCore/pull/439#issuecomment-577778145

valeriupredoi commented:
> > @ledm cheers a bunch for testing, dude! Can you pls attach the recipe you are running (the more involved one) so @mattiarighi can run a test too?
> 
> It's basically just this one, but with a couple extra SSP experiments and longer time ranges.
> 
> [#439 (comment)](https://github.com/ESMValGroup/ESMValCore/pull/439#issuecomment-577778145)

I just wrote a test with that exact recipe :beer:

valeriupredoi commented:
OK folks am gonna stop here coz I am literally fed up with this thing :grin: Functionality is good, good testing (upped to 78% avg core testing) and documentation in fair shape (maybe Lee, if you would like to add more to the doc, that'd be grand!). Fire up the review pls, I reckons this needs to get in fairly soon since it's a nice addition :beer:

ledm commented:
From my side, I'm really happy with this PR and it solved a number of significant issues with FX files in the preprocessor. I have not gone through all the code changes, so I suspect @bouweandela might catch something, but for me this PR addresses for all sorts of long-term issues. Please Merge as soon as possible!

valeriupredoi commented:
cheers @ledm -> vey happy to hear that it does the trick for you (and your heavy testing has now spared @mattiarighi from testing heh) -> @bouweandela could you pls have a looksee wt your convenience, man! Also if @schlunma has more suggs (cheers for the input mate!) :beer: all round

valeriupredoi commented:
> Great job V.!
> 
> One general thing: I would love it when the preprocessors that need fx files accepts `cube` arguments instead of filenames (or at least both). E.g. here:
> 
> https://github.com/ESMValGroup/ESMValCore/blob/bfe0db44c2c26b0b17110fea286f1d21d25ab382/esmvalcore/preprocessor/_mask.py#L89
> 
> Since most other intermediate preprocessor we try to use for fx files read cubes and output cubes, this would probably make the interface between the intermediate preprocessors and the fx file using preprocessor a lot clearer. We would also skip the unnecessary `save` and `load` step we have now.
> 
> Additionally, this would simplify the use of this preprocessor in other modules. For example, I want to use the `landsea_fraction_weighting` preprocessor in a derive script, but the derive script only has access to cubes, not filenames.

realized I didnt say anything about this - my apols man, the answer is yes but it's the same situation as when you get a bottle of beer and no glass: it would be nice to have a glass but you can drink the beer straight from the bottle for now :grin: We should get a glass in a future PR though but that will be quite a bit of work

schlunma commented:
One last comment (after that I'm happy, promised 😀): I still vote for not having to add `fx_var_preprocess` to the recipe. I think it doesn't make sense to ever set this option to `False`. I get why you need this variable when you load the ancestors, but I'm pretty sure it's possible to automatically set this to `True` when a preprocessor is used which needs fx files.

schlunma commented:
> 
> 
> > Great job V.!
> > One general thing: I would love it when the preprocessors that need fx files accepts `cube` arguments instead of filenames (or at least both). E.g. here:
> > https://github.com/ESMValGroup/ESMValCore/blob/bfe0db44c2c26b0b17110fea286f1d21d25ab382/esmvalcore/preprocessor/_mask.py#L89
> > 
> > Since most other intermediate preprocessor we try to use for fx files read cubes and output cubes, this would probably make the interface between the intermediate preprocessors and the fx file using preprocessor a lot clearer. We would also skip the unnecessary `save` and `load` step we have now.
> > Additionally, this would simplify the use of this preprocessor in other modules. For example, I want to use the `landsea_fraction_weighting` preprocessor in a derive script, but the derive script only has access to cubes, not filenames.
> 
> realized I didnt say anything about this - my apols man, the answer is yes but it's the same situation as when you get a bottle of beer and no glass: it would be nice to have a glass but you can drink the beer straight from the bottle for now 😁 We should get a glass in a future PR though but that will be quite a bit of work

One more comment: this is very easy to implement, so I'll do it myself tomorrow 👍 

E.g. this should be sufficient

```python
if not isinstance(fx_files, iris.cube.Cube):
    cube = iris.load_cube(fx_files)
else:
    cube = fx_files
```

schlunma commented on esmvalcore/_recipe.py:
```diff
@@ -365,71 +364,85 @@ def _get_default_settings(variable, config_user, derive=False):
-        if 'mip' in fx_var_dict:
-            fx_variable['mip'] = fx_var_dict['mip']
     elif fx_variable['project'] in ['OBS', 'OBS6', 'obs4mips']:
         fx_variable['mip'] = 'fx'
```
I don't think this is necessary, since the `mip` is dictated by the function calling `_add_fxvar_keys`

valeriupredoi replied:
good call, just removed it :beer:

valeriupredoi commented:
> One last comment (after that I'm happy, promised grinning): I still vote for not having to add `fx_var_preprocess` to the recipe. I think it doesn't make sense to ever set this option to `False`. I get why you need this variable when you load the ancestors, but I'm pretty sure it's possible to automatically set this to `True` when a preprocessor is used which needs fx files.

just did it man! It's all under the hood now and not changeable :beer:

valeriupredoi commented:
> > > Great job V.!
> > > One general thing: I would love it when the preprocessors that need fx files accepts `cube` arguments instead of filenames (or at least both). E.g. here:
> > > https://github.com/ESMValGroup/ESMValCore/blob/bfe0db44c2c26b0b17110fea286f1d21d25ab382/esmvalcore/preprocessor/_mask.py#L89
> > > 
> > > Since most other intermediate preprocessor we try to use for fx files read cubes and output cubes, this would probably make the interface between the intermediate preprocessors and the fx file using preprocessor a lot clearer. We would also skip the unnecessary `save` and `load` step we have now.
> > > Additionally, this would simplify the use of this preprocessor in other modules. For example, I want to use the `landsea_fraction_weighting` preprocessor in a derive script, but the derive script only has access to cubes, not filenames.
> > 
> > 
> > realized I didnt say anything about this - my apols man, the answer is yes but it's the same situation as when you get a bottle of beer and no glass: it would be nice to have a glass but you can drink the beer straight from the bottle for now grin We should get a glass in a future PR though but that will be quite a bit of work
> 
> One more comment: this is very easy to implement, so I'll do it myself tomorrow +1
> 
> E.g. this should be sufficient
> 
> ```python
> if not isinstance(fx_files, iris.cube.Cube):
>     cube = iris.load_cube(fx_files)
> else:
>     cube = fx_files
> ```

can you pls add it? Ill change the doc now and in 10min she'll be ready for you :grin:

schlunma commented:
Working on this now

valeriupredoi commented:
Awesome, Manu man! one thing I thought we should do (but not in this PR since it's alredy too much) is to extend the preprocessing capability to all other fx vars used in mask landsea and weighted masking etc - it should be just a few lines, but we gotta be careful with the case when there are no files and NE masks are applied

schlunma commented:
> Awesome, Manu man! one thing I thought we should do (but not in this PR since it's alredy too much) is to extend the preprocessing capability to all other fx vars used in mask landsea and weighted masking etc - it should be just a few lines, but we gotta be careful with the case when there are no files and NE masks are applied

Absolutely, good call! I also think that for the `sftlf`-like variables there are other preprocessors which make sense, e.g. `regrid`. I think that would partly address this issue ESMValGroup/ESMValTool#593 I opened a long time ago :grin: 

Can you open a new issue for that so we don't forget it?

valeriupredoi commented:
Nice adjustments @schlunma :beer: Am stuck in a meeting all day long today so won't be able to work but this should be ready and dusted by now - @bouweandela can you pls have a look too? Better if we get this in sooner than laterzo :grin:

schlunma commented:
I expanded the tests a little bit and realized that we haven't fully considered [this](https://github.com/ESMValGroup/ESMValCore/issues/436#issuecomment-577144414) comment of Klaus yet. For example, right now there is no check if the `regrid` preprocessor is applied to a fx file.

valeriupredoi commented:
Regrid won't fail, so I'd rather let the user choose what they want to do than imposing scientific limits to an otherwise purely mechanical framework. If they want to do bonkers stuff it's their choice. Cheers for test extension!

schlunma commented on esmvalcore/preprocessor/_volume.py:
```diff
@@ -176,16 +176,22 @@ def volume_statistics(
+            list of field:short_name for the fx variables requested or
+            list of field:dict for the fx variables requested, including
+            but not limited to: short_name, mip, experiment etc (at least
+            short_name required)
```
This is not the correct description for the `fx_files` variable used in this function, but for the `fx_files` the user has to specify in the recipe (this is changes in `_recipe.py`). Do we want to change that so that the function docstring corresponds to the actual function itself?

valeriupredoi replied:
Whats wrong with it? They list of strings or dicts

schlunma replied:
https://github.com/ESMValGroup/ESMValCore/blob/2b46632fc0d1403cc7398f7d126fd71b9d598fe7/esmvalcore/preprocessor/_volume.py#L216

No, they need to be `dict`. This is converted in `recipe.py`

valeriupredoi replied:
Ah from your latest code changes, yesh

schlunma replied:
I didn't change anything in the functions `area_statistics` and `volume_statistics` themselves :grin:

valeriupredoi replied:
Ok but then volume statistics and area statistics can simply take fx_files: ['volcello'] so they dont need to be only dicts

schlunma replied:
Yes they do need to be `dict`s. Look at the function:

https://github.com/ESMValGroup/ESMValCore/blob/2b46632fc0d1403cc7398f7d126fd71b9d598fe7/esmvalcore/preprocessor/_volume.py#L170-L224

The option you describe in the docstring are the options in the recipe. The are changed here

https://github.com/ESMValGroup/ESMValCore/blob/2b46632fc0d1403cc7398f7d126fd71b9d598fe7/esmvalcore/_recipe.py#L508-L525

to be `dict`s eventually.

valeriupredoi replied:
Aye, am on the stoopid phone so Iit's hard to see the code so I trust you :grin:

schlunma commented on esmvalcore/_recipe.py:
```diff
@@ -1275,6 +1376,11 @@ def initialize_tasks(self):
+                # preprocess the same
+                # fx var file needed by different var["activity"] but with
+                # the same output fx var file
+                _check_duplication(task)
```
Are we sure that this doesn't remove useful error messages?

valeriupredoi replied:
It does cloak the RecipeError for duplicate preprocessor files, yes, but then again it also removes the duplication itself. I reckon this way it's way more useful. We could add an extra layer of check (if fx in filename) but I dont think it's needed

schlunma commented:
I'm done now, I will do the rest of the work (allow cubes as input) in another PR, this one is already very large.

ledm commented:
I'm getting a new error, it's the same Duplicate preprocessor filename as before, but with a slightly different cause.  

```
ValueError: Duplicate preprocessor filename /home/users/ldemora/workspace/ESMValTool_private/run/esmvaltool_output/recipe_gwt_time_series_20200130_114229/preproc/time_series/areacella/CMIP6_UKESM1-0-LL_fx_piControl_r1i1p1f2_areacella.nc, please file a bug report.
```

This one occurs when you have two diagnostics that both require the same preprocessed FX file.

```
preprocessors:
    prep_sst:
        custom_order: True
        annual_statistics:
            operator: mean
        extract_region:
            start_longitude: -80.
            end_longitude: 30.
            start_latitude: -80.
            end_latitude: 80.
        area_statistics:
            operator: mean
            fx_files: [{short_name: areacello, mip: Ofx, exp: piControl, activity: CMIP}]
        regrid_time:
            frequency: yr

diagnostics:
  diag_439:
      variables:
        sst:
          short_name: tos
          preprocessor: prep_sst
          fx_var_preprocess: True
          additional_datasets:
            - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: ssp585, ensemble: r1i1p1f2, grid: gn, start_year: 2060, end_year: 2063}
        sss:
          short_name: sos
          preprocessor: prep_sst
          fx_var_preprocess: True
          additional_datasets:
            - {dataset: UKESM1-0-LL,     project: CMIP6, mip: Omon, exp: ssp585, ensemble: r1i1p1f2, grid: gn, start_year: 2060, end_year: 2063}
      scripts:
        diagnostic_439:
          script: ocean/diagnostic_timeseries.py
```

valeriupredoi commented:
looking at it right now

valeriupredoi commented:
OK that's because even if the task doesn't have duplicate filenames (`_recipe._check_duplication()` is working well) the two tasks are producing the same file; this needs to be addressed over the collective tasks and not per task; I'm getting there

valeriupredoi commented:
in the meantime @ledm if you need to run your diag quickly please put the two variables in differently named diagnostics, that'll avert this issue

ledm commented:
Hate to have more bad news, but I've also found a problem when using the preprocessors `extract_levels` and `area_statistics` together.

The error is:
```
  File "/users/modellers/ledm/workspace/ESMValTool/ESMValCore/esmvalcore/preprocessor/_regrid.py", line 430, in extract_levels
    src_levels = cube.coord(axis='z', dim_coords=True)
  File "/users/modellers/ledm/anaconda3/envs/esmvaltool/lib/python3.7/site-packages/iris/cube.py", line 1463, in coord
    raise iris.exceptions.CoordinateNotFoundError(msg)
iris.exceptions.CoordinateNotFoundError: 'Expected to find exactly 1  coordinate, but found none.'

```
It looks like the `extract_level` preprocessor is being applied to the areacello data, which doesn't have a depth dimension. 


 Try it with this recipe:

```
datasets:
  - {dataset: HadGEM2-CC, project: CMIP5, mip: Omon, exp: historical, ensemble: r1i1p1, start_year: 2001, end_year: 2004}

preprocessors:
  prep_timeseries_2: # For specific levels of 3D fields
    custom_order: true
    extract_levels:
      levels: [0., 10., 100., 1000.]
      scheme: linear_horizontal_extrapolate_vertical
    area_statistics:
      operator: mean
      fx_files: [areacello, ]
    multi_model_statistics:
      span: overlap
      statistics: [mean ]

diagnostics:

  diag_timeseries_2:
    description: Global Ocean Depth Layers mean timeseries
    variables:
      thetao: # Temperature ocean 3D
        preprocessor: prep_timeseries_2
    scripts:
      Global_Ocean_DepthLayers_mean_timeseries:
        script: ocean/diagnostic_timeseries.py

```

ledm commented:
Guessing that you know this arleady, but the recipe in [this comment](https://github.com/ESMValGroup/ESMValCore/pull/439#issuecomment-580219685) is currently putting out this error message:
```
  File "/home/users/ldemora/anaconda3_20190821/envs/esmvaltoolprivate/lib/python3.7/site-packages/iris/__init__.py", line 300, in _generate_cubes
    for cube in iris.io.load_files(part_names, callback, constraints):
  File "/home/users/ldemora/anaconda3_20190821/envs/esmvaltoolprivate/lib/python3.7/site-packages/iris/io/__init__.py", line 193, in load_files
    all_file_paths = expand_filespecs(filenames)
  File "/home/users/ldemora/anaconda3_20190821/envs/esmvaltoolprivate/lib/python3.7/site-packages/iris/io/__init__.py", line 176, in expand_filespecs
    raise IOError(msg)
OSError: One or more of the files specified did not exist:
    * "/home/users/ldemora/workspace/ESMValTool_private/run/esmvaltool_output/test_20200131_114400/preproc/diag_439/areacello/CMIP6_UKESM1-0-LL_Ofx_piControl_r1i1p1f2_areacello.nc" didn't match any files

```

valeriupredoi commented:
> Guessing that you know this arleady, but the recipe in [this comment](https://github.com/ESMValGroup/ESMValCore/pull/439#issuecomment-580219685) is currently putting out this error message:
> 
> ```
>   File "/home/users/ldemora/anaconda3_20190821/envs/esmvaltoolprivate/lib/python3.7/site-packages/iris/__init__.py", line 300, in _generate_cubes
>     for cube in iris.io.load_files(part_names, callback, constraints):
>   File "/home/users/ldemora/anaconda3_20190821/envs/esmvaltoolprivate/lib/python3.7/site-packages/iris/io/__init__.py", line 193, in load_files
>     all_file_paths = expand_filespecs(filenames)
>   File "/home/users/ldemora/anaconda3_20190821/envs/esmvaltoolprivate/lib/python3.7/site-packages/iris/io/__init__.py", line 176, in expand_filespecs
>     raise IOError(msg)
> OSError: One or more of the files specified did not exist:
>     * "/home/users/ldemora/workspace/ESMValTool_private/run/esmvaltool_output/test_20200131_114400/preproc/diag_439/areacello/CMIP6_UKESM1-0-LL_Ofx_piControl_r1i1p1f2_areacello.nc" didn't match any files
> ```


should work now, can you give it a try pls :beer:

ledm commented:
This recipe is failing:
```
# ESMValTool
# recipe_ocean_example.yml
---
documentation:
  description: |
    Recipe to demonstrate several simple plots based on the monthly ocean
    temperature. Please use this file as a template for adding additional
    fields into the ocean. This work based on the BGC-val toolkit GMD-2018-103.

  authors:
    - demora_lee

  maintainer:
    - demora_lee

  references:
    - demora2018gmd

  projects:
    - ukesm


datasets:
  - {dataset: HadGEM2-CC, project: CMIP5, mip: Omon, exp: historical, ensemble: r1i1p1, start_year: 2001, end_year: 2004}
  - {dataset: HadGEM2-ES, project: CMIP5, mip: Omon, exp: historical, ensemble: r1i1p1, start_year: 2001, end_year: 2004}

preprocessors:
  prep_timeseries_1: # For 2D fields
    custom_order: true
    area_statistics:
      operator: mean
      fx_files: [areacello, ]
    multi_model_statistics:
      span: overlap
      statistics: [mean ]

diagnostics:
  diag_timeseries_1:
    variables:
      tos: # Temperature ocean surface
        preprocessor: prep_timeseries_1
    scripts:
      Global_Ocean_Surface_mean_timeseries:
        script: ocean/diagnostic_timeseries.py
```

valeriupredoi commented:
OK guys I am seriously and honestly done with this - I wil defer any other request for a new PR once this has been merged - it is 'uuuge and really there is no room for more. There are quite a few very useful implementations in here (@ledm muchas gracias for pestering me with all the corner cases, man!) but there are some potential cases of failure that **should** be treated in new PR's. Things to be careful with:

- I have introduced two intra- and inter-tasks cleaning functions that remove the duplicate ancestors withiin a task or across asks (PreprocessorTasks) - @bouweandela and @schlunma I reckon these are very useful regardless of the special fx cases, we have more a robust set of tasks BUT please have a look at the code and see if you can spot any potential issues (Lee and myself have tested but yeah)
- the preprocessing fails if select levels is applied on fx data that doesnt have the z axis but all other preproc steps should run fine; this can be sorted out by declaring an embargoed set of preprocessing steps (eg add `exclude_preprocessors: list` as a func arg to either of the area/volume/mask preprocs that use fx files) but this in a new PR (@ledm can you pls open an issue?)

valeriupredoi commented:
PS got no idea why the tests are failing (esmpy?) - they go fine on my machine, must be the latest environment that conda is picking up, will investigate on Monday, beer o'clock now - good weekend guys! :beers:

valeriupredoi commented:
> PS got no idea why the tests are failing (esmpy?) - they go fine on my machine, must be the latest environment that conda is picking up, will investigate on Monday, beer o'clock now - good weekend guys! beers

tests fixed in #452 so don't bother :beer:

valeriupredoi commented:
ok tests fixed proper now @bouweandela can I buy you a German :beer: (or :radler: ) if you have a look sooner than later, man?

ledm commented:
What's happenning here? Has this PR stalled?

bouweandela commented:
Please have a look at this comment https://github.com/ESMValGroup/ESMValCore/issues/436#issuecomment-586365138 and let's discuss the design there, before continuing.

valeriupredoi commented:
> Please have a look at this comment [#436 (comment)](https://github.com/ESMValGroup/ESMValCore/issues/436#issuecomment-586365138) and let's discuss the design there, before continuing.

Please have a look at https://github.com/ESMValGroup/ESMValCore/issues/436#issuecomment-586366252

valeriupredoi commented:
@bouweandela let's get this one reviewed, please mate - we need these features and we need them to work proper and without having to worry about iris changes. These features work in this (rather laborious) way and have been tested quite extensively by @ledm - on top of it all this PR introduces some useful task filters too that I see useful not only for fx variables

valeriupredoi commented:
OK guys - major plot twist: while trying to fix a bug that @ledm reported yesterday I discovered that there was a major issue with the fx files generation as it was until now - for the sake of not repeating analyses, I was removing ancestors with the same name and leaving just one - but that one was totally unclear with which preprocessing set of steps was being generated with. I fixed this by:

- not removing any ancestors anymore
- writing the output preprocessed fx file in with the parent variable data
- when loading for parent variable happens, loading for the attached fx variable data happens too -> pretty much what iris should be doing (but it doesn't, really)
- fixed tests too
:beer:

valeriupredoi commented:
@ledm are you still using this branch? Can I close and nuke if not?

valeriupredoi commented:
> @ledm are you still using this branch? Can I close and nuke if not?

Frenly ping @ledm - even if you said you're still using it I'd be inclined to close it since fixing those conflicts is reason for a Nobel prize for literature

ledm commented:
Lol, this is from feb 2020? What a simple beautiful time that was! We went out for a curry in Reading - we were planning to go to Singapore. Fun times. 

I'm working on the main branch and it appears that this branch's functionality is already included there? Either that or I have some workaround in my recipes, but I had assumed that this was already long merged!

valeriupredoi commented:
ah yes, the good old pre-pandemic days, all a warm blur now :sob: 

The fx handling functionality and API has changed massively with the introduction of handling of cell measures and variables, thanks to @sloosvel we are now in the 21st century on that. I'll nuke this even though I hope this was useful to you at least before we had the current implements, so farewell a few weeks of coding and sweat :white_flag:
